src/connection.c

```diff
@@ -154,12 +155,41 @@ mpd_connection_new(const char *host, unsigned port, unsigned timeout_ms)
 						&connection->timeout,
 						&connection->error);
 		}
+	}
 #endif
 
-		if (fd == MPD_INVALID_SOCKET)
+	if (fd == MPD_INVALID_SOCKET) {
+		/* special case: try socket in XDG_RUNTIME_DIR */
+		mpd_settings_free(settings);
+		const char *xdg_runtime_dir = getenv("XDG_RUNTIME_DIR");
+		if (xdg_runtime_dir == NULL) {
+			return connection;
+		}
+		size_t xdg_socket_len = strlen(xdg_runtime_dir) + sizeof("/mpd/socket") + 1;
```

Comment:
Off-by-one bug - sizeof(string_literal) already includes the null terminator.
Also this variable's name is misleading - this isn't the length of a string, but the allocation size including the null terminator.